### PR TITLE
Set application clock speeds for p4de instances in bootstrap.sh

### PR DESCRIPTION
**Pull request by brannondorsey**
**Issue #, if available:**

Fixes https://github.com/awslabs/amazon-eks-ami/issues/954, as suggested by [hassanbabaie](https://github.com/hassanbabaie).

**Description of changes:**

Sets GPU application clocks to their maximum value for p4de.*, instances according to [this recommended value](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/optimize_gpu.html) (which is also the maximum `-ac` value for this GPU type).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**

This value has been tested on p4de.24xlarge instances in EKS. They have not been tested on p4d.* instances.

Without this change, the following errors are present in `/var/log/cloud-init-output.log` for EKS nodes using p4de.24xlarge instance types.

```
...
Treating as warning and moving on.
Enabling/disabling default auto boosted clocks is not supported for GPU: 00000000:A0:1C.0.
Treating as warning and moving on.
Enabling/disabling default auto boosted clocks is not supported for GPU: 00000000:A0:1D.0.
Treating as warning and moving on.
All done.
GPU 0: NVIDIA A100-SXM-80GB (UUID: GPU-70ba6113-4249-9a27-3c17-ffd2fb5612ac)
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:10:1C.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:10:1D.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:20:1C.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:20:1D.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:90:1C.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:90:1D.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:A0:1C.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
Specified clock combination "(MEM 1215, SM 1410)" is not supported for GPU 00000000:A0:1D.0. Run 'nvidia-smi -q -d SUPPORTED_CLOCKS' to see list of supported clock combinations
Treating as warning and moving on.
All done.
...
```

When setting application clock speeds with `nvidia-smi -ac 1593,1410` on p4de.24xlarge instances, application clock speeds are correctly set to their maximum supported value... 

```
...
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:10:1C.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:10:1D.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:20:1C.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:20:1D.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:90:1C.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:90:1D.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:A0:1C.0
Applications clocks set to "(MEM 1593, SM 1410)" for GPU 00000000:A0:1D.0
...
```

### CODE REVIEW

File: files/bootstrap.sh [hunk 1]
```diff
@@ -589,8 +589,10 @@ if command -v nvidia-smi &> /dev/null; then
     echo $GPUNAME
 
     # set application clock to maximum
```
Comment (cartermckinnon) [inline files/bootstrap.sh:591]: Can you add a comment here referencing the doc you linked?

**Comment by cartermckinnon** [comment]
Thanks for putting this together!

It would be nice to change this logic to a switch block using the instance type from IMDS, so we don't have to care about the actual nvidia chipset ID's. I'll ship it either way 👍

**Comment by amrragab8080** [comment]
We can autodiscover the max clocks from the `nvidia-smi` and set the clocks appropriately that we dont need to maintain a mapping from IMDS or if/elif statements.

File: files/bootstrap.sh (hunk 1 above)
Comment (brannondorsey) [reply files/bootstrap.sh]: Good point, addressed in https://github.com/awslabs/amazon-eks-ami/pull/1097/commits/46eb61c3dcf75f735b34f0906dad8f04aa2d4e8a.

**Comment by brannondorsey** [comment]
That's a good idea @amrragab8080. Feel free to move forward with that approach instead, now or in the future. 

Will refrain from that in this PR as this approach has been tested as is.

**Comment by brannondorsey** [comment]
> We can autodiscover the max clocks from the nvidia-smi and set the clocks appropriately that we dont need to maintain a mapping from IMDS or if/elif statements.

@amrragab8080 I've implemented your idea in https://github.com/awslabs/amazon-eks-ami/pull/1100.

**Comment by brannondorsey** [comment]
Closing this in favor of #1100.

**Pull request closed without merging**